### PR TITLE
Remove hardcoded iOS version of clang target for Mac Catalyst

### PR DESCRIPTION
**Pull request by MikeChongCan**
## Background

From `clang` 13.x, `-target x86_64-apple-ios13.0-macabi` fails while linking:

```
  = note: clang: error: invalid version number in '-target x86_64-apple-ios13.0-macabi'
```

<details>
<summary>Verbose output</summary>

```
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="[removed]" VSLANG="1033" ZERO_AR_DATE="1" "cc" "-Wl,-exported_symbols_list,/var/folders/p8/qpmzbsdn07g5gxykwfxxw7y40000gn/T/rustci8tkvp/list" "-target" "x86_64-apple-ios13.0-macabi" "/var/folders/p8/qpmzbsdn07g5gxykwfxxw7y40000gn/T/rustci8tkvp/symbols.o" "/path/to/my/[project]/[user]/target/x86_64-apple-ios-macabi/release/deps/[user].[user].a2ccc648-cgu.0.rcgu.o" "-L" "/path/to/my/[project]/[user]/target/x86_64-apple-ios-macabi/release/deps" "-L" "/path/to/my/[project]/[user]/target/release/deps" "-L" "/path/to/my/[project]/[user]/target/x86_64-apple-ios-macabi/release/build/blake3-74e6ba91506ce712/out" "-L" "/path/to/my/[project]/[user]/target/x86_64-apple-ios-macabi/release/build/blake3-74e6ba91506ce712/out" "-L" "/Users/[user]/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/x86_64-apple-ios-macabi/lib" "/var/folders/p8/qpmzbsdn07g5gxykwfxxw7y40000gn/T/rustci8tkvp/libblake3-343c1616c8f62c66.rlib" "/path/to/my/[project]/[user]/target/x86_64-apple-ios-macabi/release/deps/libcompiler_builtins-15d4f20b641cf9ef.rlib" "-framework" "Security" "-framework" "CoreFoundation" "-framework" "Security" "-liconv" "-lSystem" "-lobjc" "-framework" "Security" "-framework" "Foundation" "-lc" "-lm" "-isysroot" "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk" "-Wl,-syslibroot" "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk" "-L" "/Users/[user]/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/x86_64-apple-ios-macabi/lib" "-o" "/path/to/my/[project]/[user]/target/x86_64-apple-ios-macabi/release/deps/lib[user].dylib" "-Wl,-dead_strip" "-dynamiclib" "-Wl,-dylib" "-nodefaultlibs"
  = note: clang: error: invalid version number in '-target x86_64-apple-ios13.0-macabi'
          

warning: `[user]` (lib) generated 6 warnings
error: could not compile `[user]` due to previous error; 6 warnings emitted
```
</details>

### Minimal example

C code:

```c
#include <stdio.h>
void main() {
    int a = 1;
    int b = 2;
    int c = a + b;
    printf("%d", c);
}
```

`clang` command sample:

```
➜  202301 clang -target x86_64-apple-ios13.0-macabi main.c
clang: error: invalid version number in '-target x86_64-apple-ios13.0-macabi'
➜  202301 clang -target x86_64-apple-ios14.0-macabi main.c
main.c:2:1: warning: return type of 'main' is not 'int' [-Wmain-return-type]
void main() {
^
main.c:2:1: note: change return type to 'int'
void main() {
^~~~
int
1 warning generated.
➜  202301 clang -target x86_64-apple-ios15.0-macabi main.c
main.c:2:1: warning: return type of 'main' is not 'int' [-Wmain-return-type]
void main() {
^
main.c:2:1: note: change return type to 'int'
void main() {
^~~~
int
1 warning generated.
➜  202301 clang -target x86_64-apple-ios-macabi main.c
main.c:2:1: warning: return type of 'main' is not 'int' [-Wmain-return-type]
void main() {
^
main.c:2:1: note: change return type to 'int'
void main() {
^~~~
int
1 warning generated.

➜  202301 clang --version
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

This PR is a simplified version of #96392, inspired by https://github.com/rust-lang/cc-rs/pull/727

### CODE REVIEW

**Comment by imWildCat** [comment]
gentle @rustbot ping macos

**Comment by imWildCat** [comment]
Hello @thomcc , sorry to bother! Could you help me to ping `macos`? I saw you did this here: https://github.com/rust-lang/rust/pull/96392#issuecomment-1138890515

**Comment by thomcc** [comment]
Sure. I'll also try to take a look later (my project grant covers improving rust support for apple targets.

@rustbot ping macos

**Comment by imWildCat** [comment]
Thank you so much! No rush at all!

**Comment by imWildCat** [comment]
Hello @hkratz @inflation @nvzqz @shepmaster @thomcc,

May I have some feedbacks of this PR? Thanks in advance!

**Comment by inflation** [comment]
Do you know from which Apple LLVM version that the target triple has changed? Because I think we pass the target to the linker and this change could break on older Xcode or CLT.

**Comment by imWildCat** [comment]
> Do you know from which Apple LLVM version that the target triple has changed? Because I think we pass the target to the linker and this change could break on older Xcode or CLT.

Good question!

~~I think the first version to see this regression is Xcode with clang 13. I could only remember that I saw this since Xcode 13.2 or 13.3.
But from this list (<https://gist.github.com/yamaya/2924292>), it is 13.0.~~

~~It won't be too difficult to test supportability of these triples:~~

1. `x86_64-apple-ios13.0-macabi`
2. `x86_64-apple-ios14.0-macabi`
3. `x86_64-apple-ios-macabi`, without a specify version

Update: it won't make sense if I just talk about it here. So I did some experiments:

Sample build using macOS 11/12 with Xcode 13.0 to 14.2.0: https://github.com/imWildCat-archived/clang-triple-catalyst-test/actions/runs/4012618388/jobs/6891193915
Source: https://github.com/imWildCat-archived/clang-triple-catalyst-test/blob/main/.github/workflows/xcode.yaml
image

To sum up, catalyst triples without specifying iOS version (e.g. `x86_64-apple-ios-macabi`) can work in all version of clang.

**Comment by thomcc** [comment]
I'm inclined to suggest that it's a tier3 platform so compatibility with very old toolchains (especially at the cost of new ones) might be too much for users to expect, so we should probably just take this even if it causes issues with some old toolchains (that said it's obviously not my call to make).

That said, if someone with a wider number of XCode toolchains installed (I think @BlackHoleFox might have this setup (or did at one point), but I could also do so if they're too busy) could test on some older ones to see if this does indeed work there, that'd be useful.

**Comment by imWildCat** [comment]
Thanks for your review, @thomcc!

To address your concerns, I think we can use relatively old toolchains available on GitHub actions to see the result: https://github.com/imWildCat-archived/clang-triple-catalyst-test/actions/runs/4016588629/jobs/6899818620
(Please ignore failed tests, they are all due to unsupported triple, which is `x86_64-apple-ios13.0-macabi`)

I also want to test Xcode 11.0, which is the version Apple toolchain that supports Mac Catalyst. However, we don't have macOS 10.14 on GitHub Actions (<https://github.com/actions/runner-images/issues/2347#issuecomment-750879142>).
So I can only test Xcode 11.2.1. According to <https://gist.github.com/yamaya/2924292>, there's no much difference between the clang versions between Xcode 11.0 and Xcode 11.2.1:

```
# Xcode 11.0 (11A420a)
Apple clang version 11.0.0 (clang-1100.0.33.8)
# Xcode 11.2.1 (11B500)
Apple clang version 11.0.0 (clang-1100.0.33.12)
```

From my tests, we can see:

1. `x86_64-apple-ios13.0-macabi` is not supported anymore starting Xcode 13.3, which was released over a year ago.
    - which means Rust users on macOS with newer version of Xcode *cannot* target Mac Catalyst.
2. `x86_64-apple-ios-macabi` is supported since the very beginning. We don't need to worry about it.

**Comment by imWildCat** [comment]
@thomcc Hello! Let me add another update:
I also created a virtualized macOS 10.14, installed Xcode 11.0 (first version that can target Mac Catalyst), did the same experiment. It works with a little warning line:
```
ld: warning: OS version (5.0.0) too small, changing to 13.0.0
```

image
Screen Shot 2023-01-26 at 10 25 41 PM

File: compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub fn target() -> Target {
             // These arguments are not actually invoked - they just have
```
Comment (thomcc) [inline compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs:21]: This comment is no longer accurate. It's also not clear to me why need to change it for the `-simulator` target.

Comment (imWildCat) [reply compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs]: My apologies! It should not be in this PR coz it is for simulators.

File: compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs [hunk 2]
```diff
@@ -22,7 +22,7 @@ pub fn target() -> Target {
             // These arguments are not actually invoked - they just have
```
Comment (imWildCat) [inline compiler/rustc_target/src/spec/aarch64_apple_ios_macabi.rs:22]: not sure whether we should remove these comments?

```rs
// Taken from a clang build on Xcode 11.4.1.
// These arguments are not actually invoked - they just have
// to look right to pass App Store validation.
```

**Comment by terhechte** [comment]
`arm64-apple-ios-macabi` is the much better choice. If I remember correctly, `arm64-apple-ios13.0-macabi` was only chosen as those flag came straight out of Xcode 11. Thanks @imWildCat !

**Comment by imWildCat** [comment]
Hello @thomcc @inflation , may I have another round of review, given the conversation above?

I've got another review from @terhechte .

Thanks in advance!

**Comment by thomcc** [comment]
This looks good to me, but @wesleywiser is the assigned reviewer (I'm not a compiler reviewer).

**Comment by inflation** [comment]
LGTM

**Review by wesleywiser — APPROVED** [review]
I don't have a mac setup to test this but the testing in this thread seems sufficient for a Tier 3 target to me.

**Comment by wesleywiser** [comment]
Thanks @imWildCat! 

@bors r+
